libbpf-tools/tcpconnlat.c

```diff
@@ -0,0 +1,209 @@
+
+static struct env {
+	__u64 min_us;
+	pid_t pid;
+	bool timestamp;
+	bool verbose;
+} env;
+
+const char *argp_program_version = "tcpconnlat 0.1";
+const char *argp_program_bug_address = "<ethercflow@gmail.com>";
+const char argp_program_doc[] =
+"\nTrace TCP connects and show connection latency.\n"
+"\n"
+"USAGE: tcpconnlat [-h] [-t] [-p PID]\n"
+"\n"
+"EXAMPLES:\n"
+"    tcpconnlat              # summarize on-CPU time as a histogram"
+"    tcpconnlat 1            # trace connection latency slower than 1 ms"
+"    tcpconnlat 0.1          # trace connection latency slower than 100 us"
+"    tcpconnlat -t           # 1s summaries, milliseconds, and timestamps"
+"    tcpconnlat -p 185       # trace PID 185 only";
+
+static const struct argp_option opts[] = {
+	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help" },
+	{ "timestamp", 't', NULL, 0, "Include timestamp on output" },
+	{ "pid", 'p', "PID", 0, "Trace this PID only" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	static int pos_args;
+
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'h':
+		argp_usage(state);
+		break;
+	case 'p':
+		errno = 0;
+		env.pid = strtol(arg, NULL, 10);
+		if (errno) {
+			fprintf(stderr, "invalid PID: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 't':
+		env.timestamp = true;
+		break;
+	case ARGP_KEY_ARG:
+		if (pos_args++) {
+			fprintf(stderr,
+				"Unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		errno = 0;
+		env.min_us = strtod(arg, NULL) * 1000;
+		if (errno || env.min_us <= 0) {
+			fprintf(stderr, "Invalid delay (in us): %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		    const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+	const struct event *e = data;
+	char src[INET6_ADDRSTRLEN];
+	char dst[INET6_ADDRSTRLEN];
+	union {
+		struct in_addr  x4;
+		struct in6_addr x6;
+	} s, d;
+	static __u64 start_ts;
+
+	if (env.timestamp) {
+		if (start_ts == 0)
+			start_ts = e->ts_us;
+		printf("%-9.3f ", ((double)e->ts_us - start_ts) / 1000000);
+	}
+	if (e->af == AF_INET) {
+		s.x4.s_addr = e->saddr_v4;
+		d.x4.s_addr = e->daddr_v4;
+	} else if (e->af == AF_INET6) {
+		memcpy(&s.x6.s6_addr, &e->saddr_v6, sizeof(s.x6.s6_addr));
+		memcpy(&d.x6.s6_addr, &e->daddr_v6, sizeof(d.x6.s6_addr));
+	} else {
+		fprintf(stderr, "broken event: event->af=%d", e->af);
+		return;
+	}
+
+	printf("%-6d %-12.12s %-2d %-16s %-16s %-5d %.2f\n", e->tgid, e->comm,
+		e->af == AF_INET ? 4 : 6, inet_ntop(e->af, &s, src, sizeof(src)),
+		inet_ntop(e->af, &d, dst, sizeof(dst)), ntohs(e->dport),
+		(double)e->delta_us / 1000);
```

Comment:
same nit as above, just use double constants and compiler will do the right conversion

Reply:
Fixed, thanks!